File: frontend/src/api/plotSource.ts
import { z } from "zod";
import type { components } from "./generated/schema";
import { request } from "./client";
import { objectReferenceSchema } from "./schemas/datasets";

export type PlotSource = components["schemas"]["PlotSource"];
export const plotSourceSchema: z.ZodType<PlotSource> = z
  .object({
    schema_version: z.literal("1.0"),
    project_id: z.string(),
    plot_id: z.string(),
    version_id: z.string(),
    code: z.string().nullable().optional(),
    render_code: z.string().nullable().optional(),
    parameters: z
      .record(
        z.string(),
        z.union([z.boolean(), z.number().finite(), z.string()]),
      )
      .optional(),
    input_objects: z.array(objectReferenceSchema).optional(),
    result_bindings: z.record(z.string(), objectReferenceSchema).optional(),
    message: z.string(),
  })
  .strict();

export function getPlotSource(
  projectId: string,
  plotId: string,
  versionId: string,
) {
  return request(
    `/api/v1/projects/${encodeURIComponent(projectId)}/plots/${encodeURIComponent(plotId)}/versions/${encodeURIComponent(versionId)}/source`,
    plotSourceSchema,
  );
}

File: backend/src/vis_platform_backend/contracts/plot_source.py
from typing import Literal

from pydantic import Field

from .common import StrictModel
from .datasets import ObjectReference
from .parameters import ParameterValue


class PlotSource(StrictModel):
    schema_version: Literal["1.0"] = "1.0"
    project_id: str
    plot_id: str
    version_id: str
    code: str | None = None
    render_code: str | None = None
    parameters: dict[str, ParameterValue] = Field(default_factory=dict)
    input_objects: list[ObjectReference] = Field(default_factory=list)
    result_bindings: dict[str, ObjectReference] = Field(default_factory=dict)
    message: str
